File: bindings/python/smoke.py
from __future__ import annotations

import os

from warpigs_native import Engine, WarPigsError

library = os.environ.get("WARPIGS_NATIVE_LIB")
if not library:
    raise SystemExit("WARPIGS_NATIVE_LIB is required")

with Engine(128, 10_000, 42, library=library) as engine:
    assert engine.configuration_count == 81
    assert engine.population == 128
    config = engine.configuration(0)
    assert len(config) == 4
    assert all(char in "012" for char in config)
    assert engine.configuration_code(0) < 81
    engine.step("prepare")
    engine.step("start")
    assert engine.tick == 2
    assert engine.lifecycle(0) == 2

try:
    Engine(1_000_001, 1_000_001, 1, library=library)
except WarPigsError:
    pass
else:
    raise AssertionError("native population limit was not enforced")

print("WarPigs Python native smoke: PASS")

File: bindings/julia/smoke.jl
include("WarpPigs.jl")
using .WarpPigs

library = get(ENV, "WARPIGS_NATIVE_LIB", "")
isempty(library) && error("WARPIGS_NATIVE_LIB must point to libwarpigs_core.so")

engine = WarpPigs.Engine(library, UInt64(128), UInt64(10_000), UInt64(42))
try
    @assert WarpPigs.configuration_count(engine) == 81
    @assert WarpPigs.population(engine) == 128
    first_config = WarpPigs.configuration(engine, UInt64(0))
    @assert length(first_config) == 4
    @assert all(c -> c in ['0', '1', '2'], collect(first_config))
    WarpPigs.step!(engine, UInt32(0)) # PREPARE
    WarpPigs.step!(engine, UInt32(1)) # START
    @assert WarpPigs.tick(engine) == 2
    bins = WarpPigs.configuration_histogram(engine)
    @assert length(bins) == 81
    @assert sum(bins) == 128
finally
    WarpPigs.close(engine)
    WarpPigs.close(engine) # idempotence regression
end

println("WarPigs Julia smoke: PASS")
